Add EIP: Math checking in EVM (#6888)
ethereum/EIPs: EIPS/eip-6888.md

---
eip: 6888
title: Math checking in EVM
description: Check for math underflows overflows and division by zero at EVM level
author: Renan Rodrigues de Souza (@RenanSouza2)
discussions-to: https://ethereum-magicians.org/t/eip-math-checking/13846
status: Draft
type: Standards Track
category: Core
created: 2023-04-16
---

## Abstract

> This EIP adds many math checking to EVM arithmetic. The list includes underflows, overflows, division by zero. A new opcode is added to get the flags and clean them.

Reviewer: ```suggestion
This EIP adds many math checks to EVM arithmetic. The list includes underflows, overflows, division by zero. A new opcode is added to get the flags and clear them.
```

## Motivation

The importance of math checking in smart contract projects is very clear. It was an OpenZeppelin library and then incorporated in Solidity's default behavior. Bringing this to EVM level can combine both gas efficiency and safety.

## Specification

The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119 and RFC 8174.

> Starting from `BLOCK >= HARDFORK_BLOCK`

Author: Just checked it, it is. Paris was defined by the difficulty but Bellatrix and Shanghai were defined by timestamps

> Two new flags are added to the EVM state: unsigned error (`uerr`) and signed overferror (`serr`).

Reviewer: When adding new variables to the EVM, it can be useful to specify an existing variable that has the same scope. For example, [EIP-3074](https://eips.ethereum.org/EIPS/eip-3074)'s `authorized` has the same scope as the program counter.

From the point foward `a`, `b` and `c` references the two arguments in a math opeartion and `res` the output.

> It is define two new flags `uerr` and `serr` and both start as zero.

Reviewer: ```suggestion
It defines two new flags `uerr` and `serr` and both start as zero.
```

 - When opcode is `MUL` (`0x02`) and `(a == -1 ^ b == -(2**255)) ∨ (a == -(2**255) ^ b == -1) ∨ (a != 0 ^ abs(res) / abs(a) != abs(b))`
 - When opcode is `SDIV` (`0x05`)  or `SMOD` (`0x06`); and `b == 0 ∨ (a == -(2**255) ^ b == -1)`
 - When opcode is `SHL` (`0x1b`) and `res >> a != b` (this `>>` represents `SAR`)

The function `sgn(num)` returns the sign of the number, it can be negative, zero or positive.

A new opcode, `ARITHERR` (arithmetic error) is added, with number `0x0c`. This opcode takes 0 arguments from the stack. When executed, it pushes `2 * serr + err` and sets both flags to zero.

### gas

The gas cost for the instruction is `G_base` as it is an easily accessible information

## Rationale

EVM uses two's complement for negative numbers. The opcodes listed above triggers one or two flags depending if they are used for signed and unsigned numbers.

The test for each opcode is made with implementation friendlyness in mind. The only exception is EXP as it is hard to give a concise test as most of the others relyed on the inverse operation and there is no native `LOG`. Most `EXP` implementations will internaly use `MUL` so the flag can be drawn from that instruction.

> The flag being carryed to the stack is to make it possible to give control to the code implementation on how it will be utilized. It may perform an exceptional halting or any other behavior of it's choosing.

Reviewer: ```suggestion
The flag being carryed to the stack is to make it possible to give control to the code implementation on how it will be utilized. It may perform an exceptional halting or any other behavior of its choosing.
```

Using one operation for both flags is to save using two operations.

## Backwards Compatibility

> This EIP introduces a new opcode and changes int EVM behavior. It requires to be included in a hard fork.

Reviewer: ```suggestion
This EIP introduces a new opcode and changes in EVM behavior.
```

## Test Cases

TBD

## Reference Implementation

TBD

## Security Considerations

> This is a new EVM beahavior but each code will decide how to interact with it.

Reviewer: ```suggestion
This is a new EVM behavior but each code will decide how to interact with it.
```

---
eip: 6888
title: Math checking in EVM
description: Check for math underflows overflows and division by zero at EVM level
author: Renan Rodrigues de Souza (@RenanSouza2)
discussions-to: https://ethereum-magicians.org/t/eip-math-checking/13846
status: Draft
type: Standards Track
category: Core
created: 2023-04-16
---

## Abstract

> This EIP adds many math ckecks to EVM arithmetic. The list includes underflows, overflows, division by zero. A new opcode is added to get the flags and clear them.

Reviewer 2: ```suggestion
This EIP adds many checks to EVM arithmetic and a new opcode to get the corresponding flags and clear them. The list of check includes underflows, overflows, division by zero. 
```

## Motivation

> The importance of math checks in smart contract projects is very clear. It was an OpenZeppelin library and then incorporated in Solidity's default behavior. Bringing this to EVM level can combine both gas efficiency and safety.

Reviewer 2: ```suggestion
Its important to check some conditions in smart contracts for correctness of mathematical operations.This is primarily done in compiler tools (for e.g. earlier via an OpenZeppelin library and  now incorporated in Solidity's default behavior). Bringing these checks directly into EVM can improve both gas efficiency and safety.
```

## Specification

The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119 and RFC 8174.

Starting from `BLOCK_TIMESAMP >= HARDFORK_TIMESTAMP`

### Flags

|     Variable        | Type      | Initial Value |
| ------------------- | --------- |:------------- |
| `uerr`              | `bool`    | false         |
| `serr`              | `bool`    | false         |

> Two new flags are added to the EVM state: unsigned error (`uerr`) and signed error (`serr`). The scope of those flags are the same as the program counter. Each frame of execution has their own flags. At the frame cration they are unset and they are inaltered by any kind of call.

Reviewer 2: ```suggestion
Two new flags are added to the EVM state: unsigned error (`uerr`) and signed error (`serr`). The scope of those flags are the same as the program counter. Each frame of execution has their own flags. At the frame creation they are unset and they are updated in call.
```

> From the point foward `a`, `b` and `c` references the arguments in a math operation and `res` the output. `c` is only used if the operation takes 3 inputs.

Reviewer 3: ```suggestion
From this point forward `a`, `b` and `c` references the arguments in a math operation and `res` the output. `c` is only used if the operation takes 3 inputs.
```

Author: Fixed, thanks

 - When opcode is `SDIV` (`0x05`)  or `SMOD` (`0x06`); and `b == 0 ∨ (a == -(2**255) ^ b == -1)`
 - When opcode is `SHL` (`0x1b`) and `res >> a != b` (this `>>` represents `SAR`)

The function `sgn(num)` returns the sign of the number, it can be negative, zero or positive.

A new opcode, `ARITHERR` (arithmetic error) is added, with number `0x0c`. This opcode takes 0 arguments from the stack. When executed, it pushes `2 * serr + err` and sets both flags to zero.

### gas

The gas cost for the instruction is `G_base` as it is an easily accessible information

## Rationale

EVM uses two's complement for negative numbers. The opcodes listed above triggers one or two flags depending if they are used for signed and unsigned numbers.

The conditions described for each opcode is made with implementation friendlyness in mind. The only exception is EXP as it is hard to give a concise test as most of the others relyied on the inverse operation and there is no native `LOG`. Most `EXP` implementations will internaly use `MUL` so the flag can be drawn from that instruction.

> The flag being carryed to the stack is to make it possible to give control to the code implementation on how it will be utilized. It may perform an exceptional halting or any other behavior of its choosing.

Reviewer 2: ```suggestion
The flag being carried to the stack is to make it possible to give control to the code implementation on how it will be utilized. It may perform an exceptional halting or any other behavior of its choosing.
```

Author: I`ve uploaded a new commit, thanks for the remarks
